probe-rs/src/flashing/progress.rs

```diff
@@ -124,9 +140,18 @@ impl FlashProgress {
 pub enum ProgressEvent {
     /// The flash layout has been built and the flashing procedure was initialized.
     Initialized {
-        /// The layout of the flash contents as it will be used by the flash procedure.
+        /// Whether the chip erase feature is enabled.
+        /// If this is true, the chip will be erased before any other operation. No separate erase
+        /// progress bars are necessary in this case.
+        chip_erase: bool,
+
+        /// The layout of the flash contents as it will be used by the flash procedure, grouped by
+        /// phases (fill, erase, program sequences).
         /// This is an exact report of what the flashing procedure will do during the flashing process.
-        flash_layout: FlashLayout,
+        phases: Vec<FlashLayout>,
```

Comment:
Fill erase and program act on different units of flash. I am wondering if having the same FlashLayout type could be optimized.

Reply:
> Fill erase and program act on different units of flash.

Doesn't really matter here. We have a single entry per flashing algorithm used. Previously we derived fill/erase sizes from the flash layout, and I'm not changing that here. The important change is that we can pass multiple of these now, as opposed to one per calling FlashAlgo::program and yeeting old state.

Reply:
I did multiple (3? 4?) passes of experiments trying to come up with a FlashPlan but I discarded all that because gathering data is very hairy right now. This still shows in that StartedProgramming has a length in the event - we calculate image data way too late currently, and I'd rather not run image compression twice just for this, for example.

None of my ideas felt right, the current PR was just what seemed the simplest way to solve a few bugs here.

Reply:
Ok, thanks for the explanation :)

Reply:
So, anything you'd like me to do here?

Reply:
I think its fine for now; not yet fully sure what the MR does, need to read over it in full (reviewed 1/3 of commits) :)

Reply:
![image](https://github.com/probe-rs/probe-rs/assets/977627/533314f2-25c4-4061-9559-8158768f18aa)

If there are multiple passes, we show multiple progress bars now instead of randomly overwriting them.